### ISSUE TITLE
In a microservice context discover the available swagger-docs automatically using Eureka or Zuul

### ISSUE THREAD

**Issue by PierreBesson** [cbornet/generator-jhipster-swagger-cli#4]
As the [jhipster microservice version](https://github.com/jhipster/jhipster-experimental-microservices) is getting merged to the main generator we could provide this as an option.
However I don't know if we should discover the services directly on the Eureka registry or through a Zuul route on the gateway. The former might be good enough for a Feign client (as it need only the service ID to call it) and the latter for an Angular client (as it need to know the route to call the right REST endpoint).
What do you think @jdubois ?

Once the jhiregistry is started we can get the list of available registered microservices. Usually the registry is located on localhost:8761.

GET http://localhost:8761/eureka/apps

```
<?xml version="1.0" encoding="UTF-8"?>
<applications>
   <versions__delta>1</versions__delta>
   <apps__hashcode>UP_1_</apps__hashcode>
   <application>
      <name>APP1</name>
      <instance>
         <instanceId>app1:3bcde735f84db35caa06fd101b8d1cb0</instanceId>
         <hostName>192.168.43.8</hostName>
         <app>APP1</app>
         <ipAddr>192.168.43.8</ipAddr>
         <status>UP</status>
         <overriddenstatus>UNKNOWN</overriddenstatus>
         <port enabled="true">8090</port>
         <securePort enabled="false">443</securePort>
         <countryId>1</countryId>
         <dataCenterInfo class="com.netflix.appinfo.InstanceInfo$DefaultDataCenterInfo">
            <name>MyOwn</name>
         </dataCenterInfo>
         <leaseInfo>
            <renewalIntervalInSecs>30</renewalIntervalInSecs>
            <durationInSecs>90</durationInSecs>
            <registrationTimestamp>1454601030889</registrationTimestamp>
            <lastRenewalTimestamp>1454601451272</lastRenewalTimestamp>
            <evictionTimestamp>0</evictionTimestamp>
            <serviceUpTimestamp>1454601030889</serviceUpTimestamp>
         </leaseInfo>
         <metadata class="java.util.Collections$EmptyMap" />
         <homePageUrl>http://192.168.43.8:8090/</homePageUrl>
         <statusPageUrl>http://192.168.43.8:8080/info</statusPageUrl>
         <healthCheckUrl>http://192.168.43.8:8080/health</healthCheckUrl>
         <vipAddress>app1</vipAddress>
         <isCoordinatingDiscoveryServer>false</isCoordinatingDiscoveryServer>
         <lastUpdatedTimestamp>1454601030890</lastUpdatedTimestamp>
         <lastDirtyTimestamp>1454601030715</lastDirtyTimestamp>
         <actionType>ADDED</actionType>
      </instance>
   </application>
</applications>
```

From there it's easy to get the swagger-docs at http://192.168.43.8:8090/v2/api-docs

Or on the Zuul Gateway (on localhost:8080):

GET http://localhost:8080/routes

```
{"/app1/**":"app1"}
```

We can get the swagger-docs for app1 here : http://localhost:8080/app1/v2/api-docs

Finally, maybe after getting the list of all available endpoints on all microservices then one of the choice we could offer is to cherry-pick the ones we want to consume in our service and generate their rest client code.

**Comment by jdubois** [comment]
@PierreBesson on the master branch in generator-jhipster, generate a new "gateway", you will get a new REST endpoint provided by the GatewayResource class -> use this REST endpoint to get the URLs on which the services are mapped, and so you can then use them with Swagger

**Comment by PierreBesson** [comment]
Thanks @jdubois, I had forgotten.
To get the routes:
GET http://localhost:8080/api/gateway/routes

```
[{
    "path": "/app1/**",
    "serviceId": "app1",
    "serviceInstances": [{
        "serviceId": "APP1",
        "uri": "http://192.168.43.8:8090",
        "secure": false,
        "instanceInfo": {
            "instanceId": "app1:3bcde735f84db35caa06fd101b8d1cb0",
            "app": "APP1",
            "appGroupName": null,
            "ipAddr": "192.168.43.8",
            "sid": "na",
            "homePageUrl": "http://192.168.43.8:8090/",
            "statusPageUrl": "http://192.168.43.8:8080/info",
            "healthCheckUrl": "http://192.168.43.8:8080/health",
            "secureHealthCheckUrl": null,
            "vipAddress": "app1",
            "secureVipAddress": null,
            "countryId": 1,
            "dataCenterInfo": {
                "@class": "com.netflix.appinfo.InstanceInfo$DefaultDataCenterInfo",
                "name": "MyOwn"
            },
            "hostName": "192.168.43.8",
            "status": "UP",
            "leaseInfo": {
                "renewalIntervalInSecs": 30,
                "durationInSecs": 90,
                "registrationTimestamp": 1454601030889,
                "lastRenewalTimestamp": 1454604961768,
                "evictionTimestamp": 0,
                "serviceUpTimestamp": 1454601030889
            },
            "isCoordinatingDiscoveryServer": false,
            "metadata": {},
            "lastUpdatedTimestamp": 1454601030890,
            "lastDirtyTimestamp": 1454601030715,
            "actionType": "ADDED",
            "asgName": null,
            "overriddenStatus": "UNKNOWN"
        },
        "host": "192.168.43.8",
        "port": 8090
    }]
}]
```

**Comment by cbornet** [comment]
@PierreBesson So what would be the idea ?
The generator calls http://localhost:8080/api/gateway/routes, extracts uri for each app and proposes to generate client code for a selection of apps ?
We could have a "Select from micro-service gateway" option in the generator.

**Comment by PierreBesson** [comment]
Yes this was my idea. It would enable us to aggregate calls to multiple microservices on the gateway. Something to facilitate something like [this](http://samnewman.io/patterns/architectural/bff/#general) (for exemple).

**Comment by cbornet** [comment]
OK. Shouldn't be hard to do.

**Comment by PierreBesson** [comment]
Can you create a branch for this, I will start working on it.

**Comment by cbornet** [comment]
It's maybe better if you fork and create your own branch.
Then you can PR against master.

**Comment by cbornet** [comment]
Or we could host this project into hipster-labs ?

**Comment by cbornet** [comment]
@PierreBesson can we close from #6 ?

**Comment by PierreBesson** [comment]
Yes sure.

**Comment by cbornet** [comment]
OK. Next time you can put a "Fix #4" in your commit so that the issue is automatically closed.
